### PR TITLE
use a fixed world for code loading

### PR DESCRIPTION
**Pull request by KristofferC**
Code loading `Base.require` is one function that keeps getting run over and over as packages gets loaded. It is therefore important that this function does not get invalidated because having to recompile it causes significant latency. A lot of work has been done trying to shield it from invalidations but it seems to be a never ending battle. During the loading of OmniPackage.jl for examples, I have seen code loading gotten recompiled 4 times during the loading of that package.

A possible solution is to run the code loading code in a fixed world age thereby making it immune to invalidations (since it keeps running the old code). This also means that the code cannot be straight forwardly Revised, but maybe that is a price worth paying.

cc @c42f, @timholy, @JeffBezanson, @vtjnash

### CODE REVIEW

File: base/loading.jl [hunk 1]
```diff
@@ -1654,7 +1654,17 @@ require(uuidkey::PkgId) = @lock require_lock _require_prelocked(uuidkey)
 
 const REPL_PKGID = PkgId(UUID("3fa0cd96-eef1-5676-8a61-b3b8758bbffb"), "REPL")
 
+_require_world_age = Ref{UInt}(typemax(UInt))
```
Comment (c42f) [inline base/loading.jl:1657]:
```suggestion
const _require_world_age = Ref{UInt}(typemax(UInt))
```

**Review by c42f — APPROVED** [review]
This seems like a good tradeoff to me.

If people want to circumvent this when they're developing the loading code, can't they just do

```julia
Base._require_world_age[] = typemax(UInt)
```
?

**Comment by oscardssmith** [comment]
can this also be done for repl tab completions?

**Comment by c42f** [comment]
This can and probably should be done for any Julia "session infrastructure" or compiler-related stuff which is
* Implemented on the Julia side
* Does not need to be extended by users
* Suffers from invalidations due to package loading

REPL keybindings are an interesting case - we can't choose a fixed world for all key bindings because packages add to the keybinding mappings.  Rather, packages which extend the keybindings need to choose an appropriate world age when the keybinding is installed. For example I did this for OhMyREPL in https://github.com/KristofferC/OhMyREPL.jl/pull/321

If the REPL had an API for adding keybindings, it would make sense to capture the world age when the keybinding was added and to remove the current use of `invokelateset()` within the REPL.

**Comment by KristofferC** [comment]
After https://github.com/JuliaLang/julia/pull/49604, I am not sure I have a concrete case in the wild where this is needed.

**Comment by KristofferC** [comment]
Okay, I am seeing `require` getting invalidated again now so now I want to do this again hah.

**Comment by c42f** [comment]
I found a neat way to express the fixed world age in

https://github.com/JuliaLang/JuliaSyntax.jl/pull/290

So an option would be to steal the `fix_world_age` function from there, if it helps:

https://github.com/JuliaLang/JuliaSyntax.jl/blob/68d3fbd300fd3d03e69bfc1e99bb7eb7dcd47030/src/hooks.jl#L107

**Comment by KristofferC** [comment]
Hm, for this specific case, I don't think it brings much clarity over just doing it manually.

**Comment by KristofferC** [comment]
@nanosoldier `runtests(vs = ":master")`

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/49525#issuecomment-1566885284) has completed - possible new issues were detected.
A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/904cffd_vs_1cc10a6/report.html).

**Comment by KristofferC** [comment]
Let's try this then.

**Pull request merged**